src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -1638,14 +2061,96 @@ def __init__(self, *args, **kwargs):
             'CA_CREATE_INTERMEDIATE': self.__create_intermediate_ca,
         }
 
+    @accepts()
+    async def profiles(self):
+        """
+        Returns a dictionary of predefined options for specific use cases i.e OpenVPN certificate authority
+        configurations which can be used for creating certificate authorities.
+        """
+        return self.PROFILES
+
+    @periodic(86400, run_on_start=True)
+    @private
+    async def crl_generation(self):
+        await self.middleware.call('service.start', 'ssl')
+
     # HELPER METHODS
 
+    @private
+    async def revoke_ca_chain(self, ca_id):
+        chain = await self.get_ca_chain(ca_id)
+        for cert in chain:
+            datastore = f'system.certificate{"authority" if cert["cert_type"] == "CA" else ""}'
+            await self.middleware.call(
+                'datastore.update',
+                datastore,
+                cert['id'], {
+                    'revoked_date': datetime.datetime.utcnow()
+                },
+                {'prefix': self._config.datastore_prefix}
+            )
+
+    @private
+    async def get_ca_chain(self, ca_id):
+        certs = list(
+            map(
+                lambda item: dict(item, cert_type='CERTIFICATE'),
+                await self.middleware.call(
+                    'datastore.query',
+                    'system.certificate',
+                    [['signedby', '=', ca_id]],
+                    {'prefix': self._config.datastore_prefix}
+                )
+            )
+        )
+
+        for ca in await self.middleware.call(
+            'datastore.query',
+            'system.certificateauthority',
+            [['signedby', '=', ca_id]],
+            {'prefix': self._config.datastore_prefix}
+        ):
+            certs.extend((await self.get_ca_chain(ca['id'])))
+
+        ca = await self.middleware.call(
+            'datastore.query',
+            'system.certificateauthority',
+            [['id', '=', ca_id]],
+            {'prefix': self._config.datastore_prefix, 'get': True}
+        )
+        ca.update({'cert_type': 'CA'})
+
+        certs.append(ca)
+        return certs
+
     @private
     async def validate_common_attributes(self, data, schema_name):
         verrors = ValidationErrors()
 
         await _validate_common_attributes(self.middleware, data, verrors, schema_name)
 
+        if not data['cert_extensions']['BasicConstraints']['enabled']:
+            verrors.add(
+                f'{schema_name}.cert_extensions.BasicConstraints.enabled',
+                'For Certificate Authority, this should be enabled.'
```

Comment:
```suggestion
                'This must be enabled for a Certificate Authority'
```
Suggesting reversing the sentence order to remove the pause.